src/box/sql/vdbe.c

```diff
@@ -1968,15 +1968,27 @@ case OP_Column: {
 /* Opcode: Fetch P1 P2 P3 * *
  * Synopsis: r[P3]=PX
  *
- * Interpret data P1 points at as an initialized vdbe_field_ref
- * object.
+ * Register P1 contains an initialized vdbe_field_ref object or encoded value.
+ * In the case of a value, the register type is BIN, otherwise the type is PTR.
  *
- * If P4 is not a field name, extract the P2th field from the tuple. Otherwise,
- * get the field with the given name. The retrieved value is stored in
- * register P3.
+ * In case of P1 being encoded value, decode this value to register P3.
```

Comment:
Can you explain how P1 gets an encoded value step by step? I am struggling to find it from the code.

Reply:
I expanded the comment a bit, but I am not sure if this is enough.

Currently, the only case where this opcode can receive an encoded value is when the SQL_EXPR function is used as a field check constraint. In this case, the function gets the encoded value from the port instead. In case of tuple check constraint it takes an array containing all fields + format. I think we should rework the entire argument-taking system for SQL_EXPR functions, but I'm not sure if we should do that here, since SQL_EXPR functions currently can only work as constraints and nothing more.

Reply:
Ok, I see now. My concert is that it seems wrong to use `OP_Fetch` like that. In case of `SQL_EXPR` (field check) it doesn't really fetch anything. It only does `mem_from_mp()` on its single argument. The behaviour is too different depending on whether `p1` is bin or ptr.

What I propose below is up to you. It might simplify `OP_Fetch` (keep it like before I mean). AFAIU in `func_sql_expr_call()` you do `sql_bind_bin_static()` instead of using `vdbe_field_ref` with `sql_bind_ptr()` because  the column is not an array (so `vdbe_field_ref_prepare_data()` wouldn't work). Correct?

The proposal is to add `vdbe_field_ref_prepare_array()` which takes `uint32 field_count, const char *arr_begin, const char *arr_end`. It won't do `mp_decode_array()` inside. Then for CHECK constraints you can keep `func_sql_expr_call()` mostly like before but you call `vdbe_field_ref_prepare_array(ref, 1, data, mp_size)` for field checks. `1` here means an array with a single field. The field used in `CHECK`. We won't need `MP_ARRAY` header in the single field data for that. Tuple format won't be needed either. Because in `OP_Fetch` you will fetch by field number, not by name.

The overhead will be that we allocate `vdbe_field_ref` on region for field CHECKs, but win is that `OP_Fetch` will have less branches and code might be just simpler.

I would be glad if you could try the approach above, see if it works at all. And maybe it would be simpler.

Reply:
Thanks for the suggestion! I applied your suggestion and it made the patch simpler. However, I'm not entirely sure that `vdbe_field_ref` was designed for such usage. However, I think at some point we will change the design of `vdbe_field_ref`, for example moving it from VDBE to BOX.

Reply:
Now there is some kind of duplication. Might be better to have a single constructor for `vdbe_field_ref`, like this (I haven't tested it, and `slots[0]` makes no sense now):
<details><summary>Diff</summary>

```Diff
@@ -1374,17 +1374,14 @@ sql_template_space_new(Parse *parser, const char *name)
  */
 static void
 vdbe_field_ref_fill(struct vdbe_field_ref *field_ref, struct tuple *tuple,
-		    const char *data, uint32_t data_sz)
+		    uint32_t mp_count, const char *data, uint32_t data_sz)
 {
 	field_ref->tuple = tuple;
 	field_ref->data = data;
 	field_ref->data_sz = data_sz;
-
-	const char *field0 = data;
 	field_ref->format = NULL;
-	uint32_t mp_count = mp_decode_array(&field0);
 	field_ref->field_count = MIN(field_ref->field_capacity, mp_count);
-	field_ref->slots[0] = (uint32_t)(field0 - data);
+	field_ref->slots[0] = 0;
 	memset(&field_ref->slots[1], 0,
 	       field_ref->field_count * sizeof(field_ref->slots[0]));
 	field_ref->slot_bitmask = 0;
@@ -1395,30 +1392,28 @@ void
 vdbe_field_ref_prepare_data(struct vdbe_field_ref *field_ref, const char *data,
 			    uint32_t data_sz)
 {
-	vdbe_field_ref_fill(field_ref, NULL, data, data_sz);
+	const char *field0 = data;
+	uint32_t mp_count = mp_decode_array(&field0);
+	vdbe_field_ref_fill(field_ref, NULL, mp_count, field0,
+			    (uint32_t)(field0 - data) + data_sz);
 }
 
 void
 vdbe_field_ref_prepare_tuple(struct vdbe_field_ref *field_ref,
 			     struct tuple *tuple)
 {
-	vdbe_field_ref_fill(field_ref, tuple, tuple_data(tuple),
-			    tuple_bsize(tuple));
+	const char *field0 = tuple_data(tuple);
+	uint32_t data_sz = tuple_bsize(tuple);
+	uint32_t mp_count = mp_decode_array(&field0);
+	vdbe_field_ref_fill(field_ref, tuple, mp_count, field0,
+			    (uint32_t)(field0 - data) + data_sz);
 }
 
 void
 vdbe_field_ref_prepare_array(struct vdbe_field_ref *ref, uint32_t field_count,
 			     const char *data, uint32_t data_sz)
 {
-	ref->tuple = NULL;
-	ref->data = data;
-	ref->data_sz = data_sz;
-	ref->format = NULL;
-	ref->field_count = field_count;
-	ref->slots[0] = 0;
-	memset(&ref->slots[1], 0, ref->field_count * sizeof(ref->slots[0]));
-	ref->slot_bitmask = 0;
-	bitmask64_set_bit(&ref->slot_bitmask, 0);
+	vdbe_field_ref_fill(ref, NULL, field_count, data, data_sz);
 }
```
</details>

But I now noticed a bigger thing. Apparently another place of struggling is that you still have field names in those field-checks. That is why even in the new version you keep changing `OP_Fetch` to make it work differently depending on `format == / != NULL`. You can't properly resolve the names, because have no format. As a result we can write ridiculous stuff like below:
```Lua
box.cfg{}
def = {language = 'SQL_EXPR', is_deterministic = true, body = 'x + y == 4'}
box.schema.func.create('abc', def)
format = {{name = 'A', type = 'integer', constraint = 'abc'}}
s = box.schema.space.create('test', {format = format})
s:create_index('i')
s:replace{1} -- fails
s:replace{2} -- works
```

In the expr I used `x` and `y`, the field name was `A`, and all worked anyway. `A` was resolved to `x` and `y` at the same time. I suggest to clarify with products if this is the intended behaviour. Looks highly arguable :(

Otherwise it looks wrong tbh. We can try to think of some solution if products will confirm that this behaviour is wrong. Maybe sql_expr funcs should get complete tuple even for field checks.

Reply:
Thank you for the diff and the suggestion! As for diff - I applied it after a bit of change. As for suggestion - I discussed this with the product managers and we decided that for SQL_EXPR functions used as a field constraint from Lua, we would only check than number of arguments of the function is one. For SQL_EXPR functions used as a field constraint from SQL, we will check that the number of arguments is one and that the argument name is the same as the field name.

Also I reworked a bit `OP_Fetch` and introduced `OP_FetchByName`.

Reply:
Ok, understandable. That would allow to use the same sql checks for multiple fields.